accepted/single-file/design.md

```diff
@@ -66,225 +57,204 @@ Publishing to a single file can be triggered by adding the following property to
-  When `PublishSingleFile`  is set to `true`, it is an error to leave `RuntimeIdentifier` undefined, or to set `UseAppHost` to `false`.
-* Setting the `PublishSingleFile`property causes the managed app, managed dependencies, platform-specific native dependencies, configurations, etc. (basically the contents of the publish directory when `dotnet publish` is run without setting the property) to be embedded within the native `apphost`. 
+When the `PublishSingleFile` property is set to true, 
 
-By default, the symbol files are not embedded within the single-file, but remain as separate files in the publish directory. 
-This includes both the IL `.pdb` file, and the native `.ni.pdb` / `app.guid.map` files generated by ready-to-run compiler.
-Setting the following property causes the symbol files to be included in the single-file.
+* `RuntimeIdentifier` must be defined. Single-file builds generate a native binary for the specific platform and architecture.
+* `UseAppHost` cannot be  set to `false`.  
+* If `TargetFramework` is
+  * `netcoreapp5.0`, single-file publish works as described in this document.
+  * `netcoreapp3.0`  or `netcoreapp3.1` single-file publish works as described [here](design_3_0.md).
+  * An earlier framework, causes a compilation error.
 
-```xml
-<PropertyGroup>
-    <IncludeSymbolsInSingleFile>true</IncludeSymbolsInSingleFile>
-</PropertyGroup>
-```
+Setting the `PublishSingleFile` property causes the managed app, runtime configuration files (`app.deps.json`, `app.runtimeconfig.json`), and managed binary dependencies to be embedded within the native `apphost`.  All managed binaries (IL and ready-to-run files) that would be written to the publish directory and any sub-directories are bundled with the apphost.
 
-Certain files can be explicitly excluded from being embedded in the single-file by setting following meta-data:
+All other files, including platform-specific native binaries and symbol files, are left alongside the app by default. However, the set of files left unbundled alongside the app is expected to be small, such as: data-files (ex: `appsettings.json`) and custom native binary dependencies of the application. Further details regarding the files left next to the app is discussed in the [Host build](#Host-Builds) section. 
 
-```xml
-<ExcludeFromSingleFile>true</ExcludeFromSingleFile>
-```
+### Optional Settings
 
-For example, to place some files in the publish directory but not bundle them in the single-file:
+| Property                     | Behavior when set to `true`                                  |
+| ---------------------------- | ------------------------------------------------------------ |
+| `IncludeContentInSingleFile` | All published files (except symbol files) are embedded within the single-file bundle. The files that cannot be processed directly from the bundle are extracted out to disk at startup. |
```

Comment:
How do people find where these files are extracted to with the changed meaning of `AppContext.BaseDirectory` property?

Reply:
In my proposal I originally had a new API `GetContentRoot()` API for locating the base directory of extracted files:

```cs
namespace System.Runtime.Loader
{
    public partial class Bundle
    {
        // Check whether an app is running from a single-file bundle
        public static bool IsBundle();
        
        // Get the location where contents of the bundle are extracted, if any
        public static string GetContentRoot();
    }
}
```

However, a different viewpoint is that the extraction is somewhat an implementation detail, and the app should not learn about it. That is, the contents of the extracted directory should be accessed via internal mechanisms, For example,  native library loads through PInvoke or `NativeLibrary.Load()` API search through known paths `NATIVE_DLL_SEARCH_DIRECTORIES`.

Reply:
I would be fine with limiting the extraction to the unmanaged .dlls that cannot be embedded. In that case,  it may better to use more specific name for this, e.g. `IncludeNativeLibrariesInSingleFile`.